Move the minimum supported NumPy version to 1.21

stuartarchibald opened the pull request:
This moves the minimum supported NumPy version to 1.21. The patch removes code specialising on any version of NumPy prior to 1.21 (the oldest of which was from 1.16). It also updates code to remove workarounds necessary due to changes in the NumPy API (e.g. np.broadcast_shapes). Docs, buildscript and `setup.py`/`__init__.py` are also updated.

Closes #8707

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities here, please click the arrow besides
   "Create Pull Request" and choose "Create Draft Pull Request".
   When it's ready for review, you can click the button "ready to review" near
   the end of the pull request
   (besides "This pull request is still a work in progress".)
   The maintainers will then be automatically notified to review it.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on main/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against main they should
   be resolved by merging main into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

gmarkall commented:
gpuci run tests

gmarkall commented on numba/tests/test_np_functions.py:
```diff
@@ -2811,27 +2796,21 @@ def input_variations():
-        else:
-            # np 1.16 raises, cannot cast float64 array to intp array
-            _check_raises_type_error(params, 'to_begin')
+        # np >1.16 raises, cannot cast float64 array to intp array
```
Possibly redundant comment?

gmarkall commented on numba/tests/test_npdatetime.py:
```diff
@@ -385,12 +385,11 @@ def check(a, b, expected):
-                if np.isnat(a) or np.isnat(a):
-                    expected_val = False
-                    not_expected_val = True
+            # since np 1.16 all NaT == comparisons are False, including
```
Multiple comments contrasting NP 1.16 with older behaviour (I won't add more comments on this file to save noise)

gmarkall reviewed (commented):
I've read through the changes and they look good - the only comment I had is it looks like there are some comments that wouldn't have been written if there weren't divergent behaviour between NumPy versions - what do you think about their necessity now?

I need to go through this tomorrow to check if anything is missed, but other than that, this is looking good so far.

stuartarchibald commented on numba/np/npdatetime.py:
```diff
@@ -384,16 +384,12 @@ def impl(context, builder, sig, args):
-                if numpy_support.numpy_version < (1, 16):
-                    # No scaling when comparing NaTs
-                    builder.store(builder.icmp_unsigned(ll_op, va, vb), ret)
+                # NumPy >= 1.16 switched to NaT ==/>=/>/</<= NaT being
```
I considered deleting these comments but chose to leave them in as a reminder for anyone revisiting this code as to what the behaviour is around `NaT`... if you think it would be better/more clean to remove them am happy to do so.

gmarkall reviewed (changes requested):
I don't feel strongly about the comments.

I think this test will never run again, so could be removed:

https://github.com/numba/numba/blob/f54d55345a27197469be172a444d59384f18bf5e/numba/tests/test_np_functions.py#L2780-L2786

I also wonder whether we need to carry this macro around anymore (my assumption is that if we require NumPy 1.21 nobody should be building Numba with a version less than 1.21), but I'd maybe lean towards punting it to a separate issue / PR following this one, given it's a change at the C level and I'd rather not hold this one up:

https://github.com/numba/numba/blob/f54d55345a27197469be172a444d59384f18bf5e/numba/_numba_common.h#L24-L41

stuartarchibald commented:
Thanks for reviewing this @gmarkall.

> I don't feel strongly about the comments.

I elected to modify them, I think they are useful notes RE behaviour but the NumPy 1.16 part was/is a distraction. In the process I found another couple of redundant 1.16 references which are now removed. Commit 5d7b0fb  contains these changes.

> I think this test will never run again, so could be removed:
> 
> https://github.com/numba/numba/blob/f54d55345a27197469be172a444d59384f18bf5e/numba/tests/test_np_functions.py#L2780-L2786

Thanks for catching this, removed in dac8800.

> 
> I also wonder whether we need to carry this macro around anymore (my assumption is that if we require NumPy 1.21 nobody should be building Numba with a version less than 1.21), but I'd maybe lean towards punting it to a separate issue / PR following this one, given it's a change at the C level and I'd rather not hold this one up:
> 
> https://github.com/numba/numba/blob/f54d55345a27197469be172a444d59384f18bf5e/numba/_numba_common.h#L24-L41

I also wondered about this as it appeared when searching for references to older NumPy versions. IIRC Numba can be compiled using a much older version of NumPy than it supports at runtime, so I elected to leave it in with view of making this patch potentially a lower risk.

Finally b746d41 is a flake8 violation fix-up. Thanks again for the review.

gmarkall commented:
> I also wondered about this as it appeared when searching for references to older NumPy versions. IIRC Numba can be compiled using a much older version of NumPy than it supports at runtime, so I elected to leave it in with view of making this patch potentially a lower risk.

Sure, I didn't think it should be done in this PR. It's not clear to me why we support building with an older-than-supported NumPy  - I think that's not documented? Is it mentioned somewhere?

gmarkall reviewed (approved):
Thanks for the updates, I think this is good to merge assuming gpuCI is fine.

gmarkall commented:
gpuci run tests

stuartarchibald commented:
> > I also wondered about this as it appeared when searching for references to older NumPy versions. IIRC Numba can be compiled using a much older version of NumPy than it supports at runtime, so I elected to leave it in with view of making this patch potentially a lower risk.
> 
> Sure, I didn't think it should be done in this PR. It's not clear to me why we support building with an older-than-supported NumPy - I think that's not documented? Is it mentioned somewhere?

I'd have to go and check but I think it is/was implicit through the use of `conda build` which has/had a default NumPy version at build time. This may not be applicable or of relevance to Numba any more.
